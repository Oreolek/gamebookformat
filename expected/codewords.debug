BEGIN DEBUG OUTPUT
Book title: Gamebook
Number of sections: 9


Turn to 1 to begin.

1 (start) -  Demonstrating how [COLLECT code]Codewords[/COLLECT] (AKA sightings) can be used. Go to 2. 
2 (second) -  Got codeword [ADD code]warrior[/ADD]. Simple enough to set a codeword. Turn to 5. 
3 (xortest) -  OK, if you have the codeword [HAS code]warrior[/HAS], you may turn to 4 otherwise you may go back to 2. Although we both know you have that codeword. If you have the codeword [HAS code]fun[/HAS] you may turn to 6, without it you can go to 8. 
4 (the_end) -  That was easy.    
5 (choice) -  If you have the codeword [HAS code]warrior[/HAS] you may turn to 4. If you do not have the codeword [HAS NOT code]fun[/HAS NOT], you may turn to 8. If you have the codeword [HAS code]fun[/HAS], you may instead turn to 6. Otherwise see 3. 
6 (cheater) -  Cheater! There is no way you can have codeword fun. 
7 (noautotest) -  If you have the codeword [HAS code]fun[/HAS] turn to ((automatic))6. Otherwise you can go to the end at 4 or to the xortest at 3. 
8 (notsofun) -  This is just to demonstrate choices allowed when not having a codeword. Now go on to 9 (autotest) or 3 (xor test). 
9 (autotest) -  If you have codeword [HAS code]warrior[/HAS] turn to ((automatic))7. Otherwise you can go to the end at 4 or back to 1. 
END DEBUG OUTPUT
