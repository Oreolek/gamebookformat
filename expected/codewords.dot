digraph gamebook {

 1->2
2->9
4->2

4->5

4->8

4->3
6->4

6->3

6->5
7->6

7->3

7->1
8->4

8->7
9->3

9->8

9->5

9->4
}
