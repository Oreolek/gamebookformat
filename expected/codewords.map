{"notsofun": 8, "autotest": 7, "the_end": 3, "second": 2, "choice": 9, "xortest": 4, "start": 1, "noautotest": 6, "cheater": 5}