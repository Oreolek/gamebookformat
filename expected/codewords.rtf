{\rtf1\ansi\ansicpg1252\cocoartf1038\cocoasubrtf360
{\fonttbl\f0\fswiss\fcharset0 Helvetica;}
{\colortbl;\red255\green255\blue255;}
\paperw11900\paperh16840\margl1440\margr1440\vieww14140\viewh14860\viewkind0
\pard\tx566\tx1133\tx1700\tx2267\tx2834\tx3401\tx3968\tx4535\tx5102\tx5669\tx6236\tx6803\ql\qnatural\pardirnatural

\f0\b\fs24 \cf0

\b \qc Gamebook
\b0\

\b Turn to 1 to begin.
\b0\
\

\b \qc 1
\b0\
\ql  Demonstrating how Codewords (AKA sightings) can be used. Go to \b 2
\b0
.  \
\

\b \qc 2
\b0\
\ql  Got codeword warrior. Simple enough to set a codeword. Turn to \b 3
\b0
.  \
\

\b \qc 3
\b0\
\ql  If you have the codeword warrior you may turn to \b 9
\b0
. If you do not have the codeword fun, you may turn to \b 4
\b0
. If you have the codeword fun, you may instead turn to \b 5
\b0
. Otherwise see \b 6
\b0
.  \
\

\b \qc 4
\b0\
\ql  This is just to demonstrate choices allowed when not having a codeword. Now go on to \b 8
\b0
 (autotest) or \b 6
\b0
 (xor test).  \
\

\b \qc 5
\b0\
\ql  Cheater! There is no way you can have codeword fun.  \
\

\b \qc 6
\b0\
\ql  OK, if you have the codeword warrior, you may turn to \b 9
\b0
 otherwise you may go back to \b 2
\b0
. Although we both know you have that codeword. If you have the codeword fun you may turn to \b 5
\b0
, without it you can go to \b 4
\b0
.  \
\

\b \qc 7
\b0\
\ql  If you have the codeword fun turn to \b 5
\b0
. Otherwise you can go to the end at \b 9
\b0
 or to the xortest at \b 6
\b0
.  \
\

\b \qc 8
\b0\
\ql  If you have codeword warrior turn to \b 7
\b0
. Otherwise you can go to the end at \b 9
\b0
 or back to \b 1
\b0
.  \
\

\b \qc 9
\b0\
\ql  That was easy.     \
\

}
