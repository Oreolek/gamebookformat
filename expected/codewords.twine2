<tw-storydata name="Gamebook" startnode="1" creator="Twine" creator-version="2.0" format="Harlowe" options="">
<style role="stylesheet" id="twine-user-stylesheet" type="text/twine-css"></style><script role="script" id="twine-user-script" type="text/twine-javascript"></script>
<tw-passagedata pid="1" name="1" tags="" position="10,10"> Demonstrating how *Codewords* (AKA sightings) can be used. Go to [[2]]
. </tw-passagedata>
<tw-passagedata pid="2" name="2" tags="" position="20,20"> Got codeword *warrior*. Simple enough to set a codeword. Turn to [[9]]
. </tw-passagedata>
<tw-passagedata pid="3" name="3" tags="" position="30,30"> That was easy.    </tw-passagedata>
<tw-passagedata pid="4" name="4" tags="" position="40,40"> OK, if you have the codeword *warrior*, you may turn to [[3]]
 otherwise you may go back to [[2]]
. Although we both know you have that codeword. If you have the codeword *fun* you may turn to [[5]]
, without it you can go to [[8]]
. </tw-passagedata>
<tw-passagedata pid="5" name="5" tags="" position="50,50"> Cheater! There is no way you can have codeword fun. </tw-passagedata>
<tw-passagedata pid="6" name="6" tags="" position="60,60"> If you have the codeword *fun* turn to [[5]]
. Otherwise you can go to the end at [[3]]
 or to the xortest at [[4]]
. </tw-passagedata>
<tw-passagedata pid="7" name="7" tags="" position="70,70"> If you have codeword *warrior* turn to [[6]]
. Otherwise you can go to the end at [[3]]
 or back to [[1]]
. </tw-passagedata>
<tw-passagedata pid="8" name="8" tags="" position="80,80"> This is just to demonstrate choices allowed when not having a codeword. Now go on to [[7]]
 (autotest) or [[4]]
 (xor test). </tw-passagedata>
<tw-passagedata pid="9" name="9" tags="" position="90,90"> If you have the codeword *warrior* you may turn to [[3]]
. If you do not have the codeword *fun*, you may turn to [[8]]
. If you have the codeword *fun*, you may instead turn to [[5]]
. Otherwise see [[4]]
. </tw-passagedata>
</tw-storydata>
