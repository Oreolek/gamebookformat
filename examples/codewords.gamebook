* 1 start
Demonstrating how codewords (AKA sightings) can be used.