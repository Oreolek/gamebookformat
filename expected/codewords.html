<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta name="viewport"
        content="width=device-width, initial-scale=1, maximum-scale=1">
        <title>Gamebook</title>
    <script>


    var gamebook = {
        'player' : {
            'started' : false,
            'currentSection' : -1,
            'collections' : {},

            'collect' : function(type, name) {
                if (type in this.collections) {
                    return;
                }
                this.collections[type] = {
                    'name' : name,
                    'contents' : [],
                    'dropped' : {},
                    'add' : function(what) {
                        if (this.contents.indexOf(what) === -1
                            && !(what in this.dropped)) {
                            this.contents.push(what);
                            this.contents.sort();
                        }
                    },
                    'drop' : function(what) {
                        var i = this.contents.indexOf(what);
                        if (i >= 0) {
                            this.contents.splice(i, 1);
                            this.dropped[what] = true;
                        }
                    },
                    'has' : function(what) {
                        return this.contents.indexOf(what) >= 0;
                    }
                };
                gamebook.addCollectionView(type, name);
            },

            'add' : function(type, what) {
                this.collections[type].add(what);
                gamebook.updateCollectionsView();
            },

            'drop' : function(type, what) {
                this.collections[type].drop(what);
                gamebook.updateCollectionsView();
            },

            'has' : function(type, what) {
                return this.collections[type].has(what);
            },

            'getState' : function() {
                return JSON.stringify({
                    'collections' : this.collections,
                    'currentSection' : this.currentSection
                });
            },

            'setState' : function(state) {
                var parsedState = JSON.parse(state);
                this.currentSection = parsedState.currentSection;
                for (var c in parsedState.collections) {
                    var collection = parsedState.collections[c];
                    if (!(c in this.collections)) {
                        this.collect(c, collection.name);
                    } else {
                        this.collections[c].name = collection.name;
                    }
                    this.collections[c].contents = collection.contents;
                    this.collections[c].dropped = collection.dropped;
                }
            }
        },

        'sections' : {},

        'turnToFunctions' : {},

        'addSection' : function(nr, element) {
            var section = {'element' : element, 'nr' : nr};
            this.sections[nr] = section;
        },

        'turnTo' : function(nr) {
            if (!gamebook.player.started) {
                gamebook.start();
            }
            if (!nr in this.sections) {
                throw new Exception("Can not turn to non-existing section "
                                    + nr + ".");
            }
            this.displaySection(nr);
        },

        'start' : function() {
            this.hideIntroSections();
            this.addClassToClass('startlink', 'nodisplay');
            this.addClassToClass('resumelink', 'nodisplay');
            gamebook.player.started = true;
        },

        'displaySection' : function(nr) {
            if (this.player.currentSection > 0) {
                var section = this.sections[this.player.currentSection];
                section.element.style.display = 'none';
            }
            var e = this.sections[nr].element;
            this.runActions(e.getElementsByClassName('sectiontext')[0]);
            e.style.display = 'block';
            this.player.currentSection = nr;
            this.saveGame();
        },

        //FIXME move out from gamebook object
        'saveGame' : function() {
            if (typeof window !== 'undefined' && 'localStorage' in window) {
                window.localStorage.setItem('savedGamebookPlayer',
                                            this.player.getState());
            }
        },

        //FIXME move out from gamebook object
        'hasSavedGame' : function() {
            if (typeof window !== 'undefined' && 'localStorage' in window) {
                return window.localStorage.getItem('savedGamebookPlayer');
            } else {
                return false;
            }
        },

        //FIXME move out from gamebook object
        'loadGame' : function() {
            if (typeof window !== 'undefined' && 'localStorage' in window) {
                var state = window.localStorage.getItem('savedGamebookPlayer');
                this.player.setState(state);
                this.turnTo(this.player.currentSection);
                this.updateCollectionsView();
            } else {
                //FIXME some kind of error, because we should never get here
            }
        },

        'hideIntroSections' : function() {
            this.addClassToClass('introsection', 'nodisplay');
            this.removeClassFromClass('displayintrolink', 'nodisplay');
            this.addClassToClass('hideintrolink', 'nodisplay');
        },

        'showIntroSections' : function() {
            this.runActionsInIntroSections();
            this.removeClassFromClass('introsection', 'nodisplay');
            this.addClassToClass('displayintrolink', 'nodisplay');
            this.removeClassFromClass('hideintrolink', 'nodisplay');
            document.body.scrollIntoView();
        },

        'runActionsInIntroSections' : function() {
            Array.prototype.forEach.call(
                document.getElementsByClassName('introsectionbody'),
                gamebook.runActions);
        },

        'addClassToClass' : function(className, addClass) {
            Array.prototype.forEach.call(
                document.getElementsByClassName(className),
                function(e) {
                    e.classList.add(addClass);
                });
        },

        'removeClassFromClass' : function(className, removeClass) {
            Array.prototype.forEach.call(
                document.getElementsByClassName(className),
                function(e) {
                    e.classList.remove(removeClass);
                });
        },

        'runActions' : function(e) {
            var enableNextLink = true;
            var hasXorScope = false;
            var hasAutoScope = false;
            var xorEnableNext = false;
            var autoDisableAllRemainingLinks = (
                gamebook.player.started && e.classList.contains('introsectionbody'));
            Array.prototype.forEach.call(e.childNodes, function(c) {
                if (!c.classList) {
                    return;
                }
                if (c.classList.contains('sectionref')) {
                    if (enableNextLink && !autoDisableAllRemainingLinks) {
                        gamebook.enableLink(c);
                        if (hasAutoScope) {
                            autoDisableAllRemainingLinks = true;
                        }
                    } else {
                        gamebook.disableLink(c);
                    }
                    enableNextLink = !(hasXorScope && !xorEnableNext);
                    hasAutoScope = false;
                    hasXorScope = false;
                } else if (c.classList.contains('collect')) {
                    gamebook.player.collect(c.dataset.type, c.dataset.name);
                } else if (c.classList.contains('add')) {
                    gamebook.player.add(c.dataset.type, c.dataset.what);
                } else if (c.classList.contains('drop')) {
                    gamebook.player.drop(c.dataset.type, c.dataset.what);
                } else if (c.classList.contains('has')) {
                    enableNextLink = gamebook.player.has(c.dataset.type,
                                                         c.dataset.what);
                } else if (c.classList.contains('hasnot')) {
                    enableNextLink = !gamebook.player.has(c.dataset.type,
                                                          c.dataset.what);
                } else if (c.classList.contains('xor')) {
                    hasXorScope = true;
                    xorEnableNext = !enableNextLink;
                } else if (c.classList.contains('auto')) {
                    hasAutoScope = true;
                } else if (c.classList.contains('random')) {
                    c.addEventListener('click',
                                       gamebook.enableRandomLinkAfter);
                    c.classList.add("enabledlink");
                    c.classList.remove("disabledlink");
                    autoDisableAllRemainingLinks = true;
                }
            });
        },

        'enableLink' : function(e) {
            e.addEventListener('click',
                               gamebook.getTurnToFunction(e.dataset.ref));
            e.classList.add("enabledlink");
            e.classList.remove("disabledlink");
        },

        'disableLink' : function(e) {
            e.removeEventListener('click',
                               gamebook.getTurnToFunction(e.dataset.ref));
            e.classList.remove("enabledlink");
            e.classList.add("disabledlink");
        },

        'enableRandomLinkAfter' : function(event) {
            this.classList.remove("enabledlink");
            this.classList.add("disabledlink");
            var links = [];
            var e = this.nextSibling;
            while (e) {
                if (e.classList && e.classList.contains('sectionref')) {
                    links.push(e);
                }
                e = e.nextSibling;
            }
            if (links.length > 0) {
                var selected = links[Math.floor(Math.random()*links.length)]
                gamebook.enableLink(selected);
            } else {
                console.log("Random with nothing to select?");
            }
            event.preventDefault();
        },

        'addCollectionView' : function(type, name) {
            var ce = document.getElementById('collections');
            var template = document.getElementById('collectionTemplate');
            var e = template.cloneNode(true);
            e.className = "collection";
            e.getElementsByClassName('collectionheading')[0].innerHTML = name;
            e.dataset.type = type;
            ce.appendChild(e);
        },

        'updateCollectionsView' : function() {
            var ce = document.getElementById('collections');
            Array.prototype.forEach.call(ce.childNodes, function(c) {
                if (c.className === 'collection') {
                    var type = c.dataset.type;
                    var collection = gamebook.player.collections[type];
                    var cc = c.getElementsByClassName('collectioncontents')[0];
                    cc.innerHTML = collection.contents.join(', ');
                }
            });
        },

        'getTurnToFunction' : function(nr) {
            if (nr in this.turnToFunctions) {
                return this.turnToFunctions[nr];
            } else {
                var f = function () {
                    gamebook.turnTo(nr);
                };
                this.turnToFunctions[nr] = f;
                return f;
            }
        }

    };

// little hack to make easy to test from node.js
if (typeof exports !== 'undefined') {
    exports.gamebook = gamebook;
}


    </script>
  <style>
    .startlink,.sectionref,.displayintrolink,.hideintrolink,.found,.add,.random,
    .resumelink{
                   font-weight: bold;
                   margin: 0.2em;
                   vertical-align: middle;
                   white-space: nowrap;
    }

    .sectionref.enabledlink {
        padding-left: 1.5em;
        padding-right: 1.5em;
    }

    .startlink,.enabledlink,.displayintrolink,.hideintrolink,.found,.random,
    .resumelink {cursor: pointer;}

    .startlink,.enabledlink,.displayintrolink,.hideintrolink,.found,
    .random,.resumelink {background: #eef;
    }
    .startlink:hover,.enabledlink:hover,.displayintrolink:hover,
       .hideintrolink:hover,.resumelink:hover {background: #a9f;}
    .disabledlink {color: #bbb; cursor: not-allowed; background: inherit;}
    .disabledlink:hover {background: inherit;}
    .sectionnumber {font-weight: bolder;
                    margin-left: 50%;
                    margin-right: 50%;}
    .section {display: none; width: 90%; margin-left: 5%; margin-right: 5%;}
    .sectiontext {margin-top: 0.5em;}
    .gamebook {max-width: 30em; padding: 1em; width: 100%; font-size: 133%;}
    .collections {margin-top: 4em;}
    .collection {background: #ddd; margin-top: 1em;}
    .collectionheading {}
    .collectionheading::after {content: ": ";}
    .collectioncontents {}
    .collect {}
    .add {font-weight: bold;}
    .drop {font-style: italic;}
    .has {font-style: italic;}
    .hasnot {font-style: italic;}
    .collectionTemplate {display: none;}
    .sectionimage {width: 100%; padding: 1em;}
    .nodisplay {display: none;}
    .resumenr {font-weight: bold;}
  </style>
  </head>
  <body>
<div class="hideintrolink nodisplay"
    onclick="gamebook.hideIntroSections()">(hide instructions)</div>
  <div class="gamebook">

<div class="resumelink nodisplay"
  onclick="gamebook.loadGame()">Resume saved game.</div>

<div class="startlink"
  onclick="gamebook.turnTo(1)">Turn to 1 to begin.</div>
<script>
    if (gamebook.hasSavedGame()) {
        var resumeLinks = document.getElementsByClassName('resumelink');
        Array.prototype.forEach.call(resumeLinks, function(e) {
            e.classList.remove('nodisplay');
        });
    }
</script>
<div class="section" id="section1">
<span class="sectionnumber" id="para1">1</span>
<div class="sectiontext">
 Demonstrating how <span class="collect" data-type="code"
    data-name="Codewords">Codewords</span> (AKA sightings) can be used. Go to <a class="sectionref enabledlink" data-ref="2">2</a>. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(1, document.getElementById('section1'));
}
</script><div class="section" id="section2">
<span class="sectionnumber" id="para2">2</span>
<div class="sectiontext">
 Got codeword <span class="add" data-type="code"
      data-what="warrior">warrior</span>. Simple enough to set a codeword. Turn to <a class="sectionref enabledlink" data-ref="9">9</a>. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(2, document.getElementById('section2'));
}
</script><div class="section" id="section3">
<span class="sectionnumber" id="para3">3</span>
<div class="sectiontext">
 That was easy.    
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(3, document.getElementById('section3'));
}
</script><div class="section" id="section4">
<span class="sectionnumber" id="para4">4</span>
<div class="sectiontext">
 OK, if you have the codeword <span class="has" data-type="code" data-what="warrior">warrior</span>, you may turn to <span class="xor"></span><a class="sectionref enabledlink" data-ref="3">3</a> otherwise you may go back to <a class="sectionref enabledlink" data-ref="2">2</a>. Although we both know you have that codeword. If you have the codeword <span class="has" data-type="code" data-what="fun">fun</span> you may turn to <span class="xor"></span><a class="sectionref enabledlink" data-ref="5">5</a>, without it you can go to <a class="sectionref enabledlink" data-ref="8">8</a>. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(4, document.getElementById('section4'));
}
</script><div class="section" id="section5">
<span class="sectionnumber" id="para5">5</span>
<div class="sectiontext">
 Cheater! There is no way you can have codeword fun. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(5, document.getElementById('section5'));
}
</script><div class="section" id="section6">
<span class="sectionnumber" id="para6">6</span>
<div class="sectiontext">
 If you have the codeword <span class="has" data-type="code" data-what="fun">fun</span> turn to <span class="auto"></span><a class="sectionref enabledlink" data-ref="5">5</a>. Otherwise you can go to the end at <a class="sectionref enabledlink" data-ref="3">3</a> or to the xortest at <a class="sectionref enabledlink" data-ref="4">4</a>. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(6, document.getElementById('section6'));
}
</script><div class="section" id="section7">
<span class="sectionnumber" id="para7">7</span>
<div class="sectiontext">
 If you have codeword <span class="has" data-type="code" data-what="warrior">warrior</span> turn to <span class="auto"></span><a class="sectionref enabledlink" data-ref="6">6</a>. Otherwise you can go to the end at <a class="sectionref enabledlink" data-ref="3">3</a> or back to <a class="sectionref enabledlink" data-ref="1">1</a>. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(7, document.getElementById('section7'));
}
</script><div class="section" id="section8">
<span class="sectionnumber" id="para8">8</span>
<div class="sectiontext">
 This is just to demonstrate choices allowed when not having a codeword. Now go on to <a class="sectionref enabledlink" data-ref="7">7</a> (autotest) or <a class="sectionref enabledlink" data-ref="4">4</a> (xor test). 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(8, document.getElementById('section8'));
}
</script><div class="section" id="section9">
<span class="sectionnumber" id="para9">9</span>
<div class="sectiontext">
 If you have the codeword <span class="has" data-type="code" data-what="warrior">warrior</span> you may turn to <a class="sectionref enabledlink" data-ref="3">3</a>. If you do not have the codeword <span class="hasnot" data-type="code"
      data-what="fun">fun</span>, you may turn to <a class="sectionref enabledlink" data-ref="8">8</a>. If you have the codeword <span class="has" data-type="code" data-what="fun">fun</span>, you may instead turn to <a class="sectionref enabledlink" data-ref="5">5</a>. Otherwise see <a class="sectionref enabledlink" data-ref="4">4</a>. 
</div>
</div>
<script>
if (this.gamebook) {
   gamebook.addSection(9, document.getElementById('section9'));
}
</script>     <div id="collections" class="collections">
    </div>
    <div id="collectionTemplate" class="collectionTemplate">
      <span class="collectionheading"></span>
      <span class="collectioncontents"></span>
    </div>
    </div>
    <script>
if (this.gamebook) {
    gamebook.runActionsInIntroSections();
}
    </script>
<div class="displayintrolink nodisplay"
    onclick="gamebook.showIntroSections()">(show instructions)</div>
    </div>
  </body>
</html>
